File: frontend/src/utils/logger.ts
/**
 * 日志工具 - 捕获应用日志并发送到后端
 */

interface LogEntry {
  timestamp: number
  level: 'info' | 'warn' | 'error' | 'debug'
  message: string
  details?: any
}

class Logger {
  private logs: LogEntry[] = []
  private maxLogs = 500
  private originalConsole = {
    log: console.log,
    warn: console.warn,
    error: console.error
  }

  constructor() {
    this.interceptConsole()
    this.loadStoredLogs()
  }

  private interceptConsole() {
    // 拦截 console.log
    console.log = (...args: any[]) => {
      this.originalConsole.log(...args)
      this.addLog('info', this.formatArgs(args))
    }

    // 拦截 console.warn
    console.warn = (...args: any[]) => {
      this.originalConsole.warn(...args)
      this.addLog('warn', this.formatArgs(args))
    }

    // 拦截 console.error
    console.error = (...args: any[]) => {
      this.originalConsole.error(...args)
      this.addLog('error', this.formatArgs(args))
    }
  }

  private formatArgs(args: any[]): string {
    return args.map(arg => {
      if (typeof arg === 'object') {
        try {
          return JSON.stringify(arg)
        } catch {
          return String(arg)
        }
      }
      return String(arg)
    }).join(' ')
  }

  private addLog(level: LogEntry['level'], message: string, details?: any) {
    const entry: LogEntry = {
      timestamp: Date.now(),
      level,
      message: message.substring(0, 500), // 限制长度
      details
    }

    this.logs.push(entry)

    // 限制日志数量
    if (this.logs.length > this.maxLogs) {
      this.logs = this.logs.slice(-this.maxLogs)
    }

    // 保存到 localStorage
    this.saveLogs()
  }

  private saveLogs() {
    try {
      localStorage.setItem('app_logs', JSON.stringify(this.logs))
    } catch (e) {
      // localStorage 可能已满
    }
  }

  private loadStoredLogs() {
    try {
      const stored = localStorage.getItem('app_logs')
      if (stored) {
        this.logs = JSON.parse(stored)
      }
    } catch (e) {
      this.logs = []
    }
  }

  // 公共方法
  info(message: string, details?: any) {
    this.addLog('info', message, details)
  }

  warn(message: string, details?: any) {
    this.addLog('warn', message, details)
  }

  error(message: string, details?: any) {
    this.addLog('error', message, details)
  }

  debug(message: string, details?: any) {
    this.addLog('debug', message, details)
  }

  getLogs(): LogEntry[] {
    return [...this.logs]
  }

  clearLogs() {
    this.logs = []
    localStorage.removeItem('app_logs')
  }
}

// 导出单例
export const logger = new Logger()
export default logger

File: backend/app/core/logger.py
"""
日志工具模块
"""
import time
from datetime import datetime
from typing import List, Dict, Any, Optional

_logs: List[Dict[str, Any]] = []
_max_logs = 5000


def _cleanup_old_logs():
    """清理超过 7 天的日志"""
    global _logs
    seven_days_ago = int((time.time() - 7 * 24 * 60 * 60) * 1000)
    _logs = [log for log in _logs if log.get("timestamp", 0) > seven_days_ago]


def add_log(level: str, message: str, details: Any = None, user: Optional[str] = None, ip: Optional[str] = None):
    """添加日志"""
    global _logs
    log_entry = {
        "timestamp": int(time.time() * 1000),
        "datetime": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
        "level": level.lower(),
        "message": message,
        "details": details,
        "user": user,
        "ip": ip
    }
    _logs.insert(0, log_entry)
    _cleanup_old_logs()
    if len(_logs) > _max_logs:
        _logs = _logs[:_max_logs]


def get_logs(limit: int = 100, level: Optional[str] = None) -> List[Dict[str, Any]]:
    """获取日志"""
    global _logs
    logs = _logs[:limit]
    if level:
        logs = [log for log in logs if log.get("level") == level.lower()]
    return logs


def clear_logs():
    """清空日志"""
    global _logs
    _logs = []
